server-utils/src/suspendable_stream.rs

```diff
@@ -0,0 +1,88 @@
+use std::time::{Duration, Instant};
+use tokio::timer::Delay;
+use std::io;
+use tokio::prelude::*;
+
+/// `Incoming` is a stream of incoming sockets
+/// Polling the stream may return a temporary io::Error (for instance if we can't open the connection because of "too many open files" limit)
+/// we use for_each combinator which:
+/// 1. Runs for every Ok(socket)
+/// 2. Stops on the FIRST Err()
+/// So any temporary io::Error will cause the entire server to terminate.
+/// This wrapper type for tokio::Incoming stops accepting new connections
+/// for a specified amount of time once an io::Error is encountered
+pub struct SuspendableStream<S> {
+	stream: S,
+	next_delay: Duration,
+	initial_delay: Duration,
+	max_delay: Duration,
+	timeout: Option<Delay>,
+}
+
+impl<S> SuspendableStream<S> {
+	/// construct a new Suspendable stream, given tokio::Incoming
+	/// and the amount of time to pause for.
+	pub fn new(stream: S) -> Self {
+		SuspendableStream {
+			stream,
+			next_delay: Duration::from_millis(20),
+			initial_delay: Duration::from_millis(10),
+			max_delay: Duration::from_secs(5),
+			timeout: None,
+		}
+	}
+}
+
+impl<S, I> Stream for SuspendableStream<S>
+	where S: Stream<Item=I, Error=io::Error>
+{
+	type Item = I;
+	type Error = ();
+
+	fn poll(&mut self) -> Result<Async<Option<Self::Item>>, ()> {
+		if let Some(mut timeout) = self.timeout.take() {
+			match timeout.poll() {
+				Ok(Async::Ready(_)) => {}
+				Ok(Async::NotReady) => {
+					self.timeout = Some(timeout);
+					return Ok(Async::NotReady);
+				}
+				Err(_) => unreachable!("Polling a delay shouldn't yield any errors; qed")
```

Comment:
That is not correct, please look into my last comment here: https://github.com/paritytech/jsonrpc/pull/321#discussion_r222928243